nft-client/src/cts1.rs

```diff
@@ -0,0 +1,377 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", hex::encode(&self.0))?;
+        Ok(())
+    }
+}
+
+/// Serialize the token ID according to CTS1 specification.
+impl Serial for TokenIdVec {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        let len = u8::try_from(self.0.len()).map_err(|_| W::Err::default())?;
+        len.serial(out)?;
+        for byte in &self.0 {
+            byte.serial(out)?;
+        }
+        Ok(())
+    }
+}
+
+/// Deserialize bytes to a Token ID according to CTS1 specification.
+impl Deserial for TokenIdVec {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let tokens_id_length = u8::deserial(source)?;
+        let mut bytes = Vec::with_capacity(tokens_id_length.into());
+        for _ in 0..tokens_id_length {
+            let byte = source.read_u8()?;
+            bytes.push(byte);
+        }
+        Ok(TokenIdVec(bytes))
+    }
+}
+
+/// The parameter for the NFT Contract function "CTS1-NFT.mint".
+/// Important: this is specific to this NFT smart contract and contract
+/// functions for minting is not part of the CTS1 specification.
+#[derive(Debug)]
+pub struct MintParams {
+    pub owner:     Address,
+    pub token_ids: Vec<TokenIdVec>,
+}
+
+/// Serialization for the minting contract function parameter.
+/// Must match the serialization specified in the NFT smart contract.
+impl Serial for MintParams {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        self.owner.serial(out)?;
+        let len = u8::try_from(self.token_ids.len()).map_err(|_| W::Err::default())?;
+        len.serial(out)?;
+        for token in &self.token_ids {
+            token.serial(out)?;
+        }
+        Ok(())
+    }
+}
+
+/// Additional data bytes included with transfers the parameter for the CTS1
+/// contract function "transfer".
+#[derive(Debug, Clone)]
+pub struct ReceiveHookData(pub Vec<u8>);
+
+/// Serialization for the additional data, serialized as according to the CTS1
+/// specification.
+impl Serial for ReceiveHookData {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        let len = u16::try_from(self.0.len()).map_err(|_| W::Err::default())?;
+        len.serial(out)?;
+        for byte in &self.0 {
+            byte.serial(out)?;
+        }
+        Ok(())
+    }
+}
+
+/// Parse the additional data from a hex string.
+impl FromStr for ReceiveHookData {
+    type Err = hex::FromHexError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(ReceiveHookData(hex::decode(s)?.to_vec()))
+    }
+}
+
+/// A description of a transfer as according to the CTS1 specification.
+#[derive(Debug)]
+pub struct Transfer {
+    pub token_id:     TokenIdVec,
+    pub amount:       u64,
+    pub from:         Address,
+    pub to:           Address,
+    pub receive_name: Option<OwnedReceiveName>,
+    pub data:         ReceiveHookData,
+}
```

Comment:
Please document a bit. In particular the receive_name, why is it an option?

Reply:
I would also recommend combining the to and receive_name fields, since receive_name is only relevant if the to address is a contract, so this type has too much freedom.